flips/sync-protocol.md

```diff
@@ -0,0 +1,90 @@
+
+## Proposal
+
+The `RequestBlock` API should be updated to accept a block height, which should be stored with the queued item in `BlockIDs`. This will allow items in `BlockIDs` to be pruned as soon as the local finalized height surpasses their associated block heights. This also allows the Sync Core to ignore calls to `RequestBlock` for block heights which exceed the local finalized height by more than a configurable threshold. This helps to reduce the amount of resources spent tracking and requesting blocks which which cannot immediately be finalized anyways.
+
+Instead of pruning on every call to `ScanPending`, the Sync Core should keep track of the local finalized height from the latest call to `ScanPending`, and only trigger pruning if the height has actually changed. If necessary, it's possible to optimize the performance of pruning and avoid iterating through every item in `BlockIDs` by maintaining an additional mapping from block heights to the set of requestable block IDs at each height. 
+
+We should include a nonce with each Batch / Range Request, which replyers must include in their Block Responses. This nonce can be used to perform some validations:
+* The nonce in the response should correspond to an actual request that was sent.
+* If the response is for a Range Request, it should contain a single chain of blocks which begins at the start height of the requested range and is no longer than the size of the requested range.
+* If the response is for a Batch Request, it should contain a subset of the requested block IDs.
+
+Since the nonce allows us to distinguish between responses to Batch and Range Requests, we can avoid optimistically setting the statuses of heights as Received for responses to Batch Requests.
+
+At any time, there is a single range of heights that the Sync Engine actively requests, which is tracked by the Sync Core. We call this the Active Range. The Active Range is parameterized by two variables, which effectively replace the `Heights` map from the existing implementation:
+* `RangeStart` is the start height of the range, which is greater than or equal to `LocalFinalizedHeight`, the local finalized block height.
+* `RangeEnd` is the end height of the range, which is less than or equal to `TargetFinalizedHeight`, the speculated finalized block height of the overall chain (more details below).
+
+The Active Range has a default size of `DefaultRangeSize`. `RangeStart` is updated according to the following conditions, and `RangeEnd` should always be updated as necessary to satisfy `RangeEnd - RangeStart >= DefaultRangeSize`:
+* If the local finalized height has increased by `h`, increase `RangeStart` by `h`.
+* If every height in the Active Range has been received from at least `MinResponses` different nodes, increase `RangeStart` by `DefaultRangeSize`.
```

Comment:
```suggestion
* If every height in the Active Range has been received from at least `MinResponses` different nodes, set `RangeStart=RangeEnd+1`.
```
Since we allow for the Active Range to be `>=DefaultRangeSize`